test/test_tensor_creation_ops.py

```diff
@@ -1960,6 +1960,88 @@ def test_inference(default_dtype):
+                stack.append([np.random.randn(100, 1).astype(dtype) for _ in range(repeats)])
+                stack.append([np.random.randn(10, 8, 7).astype(dtype) for _ in range(repeats)])
+
+                for arrays in stack:
+                    self.assertEqual(torch.tensor(arrays, device=device), torch.from_numpy(np.stack(arrays)),
+                                     exact_dtype=True, rtol=1e-05, atol=1e-08)
+
+
+        # Extended tests
+        for dtype in np_types:
+            for _ in range(repeats):
+                stack = []
+                A = np.random.randn(100).astype(dtype)
+                stack.append([[A, 2 * A], [A, A]])
+                A = np.random.randn(10, 5).astype(dtype)
+                stack.append([[A, 2 * A], [A, A]])
+                A = np.random.randn(10, 10).astype(dtype)
+                stack.append([[A.T, A], [A, A.T]])
+                A = np.random.randn(20, 20).astype(dtype)
+                stack.append([[A[5:10, :], A.T[11:16, :]], [A[:5, :], A[15:20, :]]])
+                A = np.random.randn(20, 20, 20).astype(dtype)
+                stack.append([[A[5:10, :, :], A.T[11:16, :, :]], [A[:5, :, :], A[15:20, :, :]]])
+                stack.append([[[[A]]]])
+
+                for arrays in stack:
+                    self.assertEqual(torch.tensor(arrays, device=device), torch.from_numpy(np.stack(arrays)),
+                                     exact_dtype=True, rtol=1e-05, atol=1e-08)
+
+
+        # Extended tests including fortran layout arrays
+        for dtype in np_types:
+            for _ in range(repeats):
+                stack = []
+                A = np.asfortranarray(np.random.randn(100).astype(dtype))
+                stack.append([[A, 2 * A], [A, A]])
+                A = np.asfortranarray(np.random.randn(10, 5).astype(dtype))
+                stack.append([[A, 2 * A], [A, A]])
+                A = np.asfortranarray(np.random.randn(10, 10).astype(dtype))
+                stack.append([[A.T, A], [A, A.T]])
+                A = np.asfortranarray(np.random.randn(20, 20).astype(dtype))
+                stack.append([[A[5:10, :], A.T[11:16, :]], [A[:5, :], A[15:20, :]]])
+                A = np.asfortranarray(np.random.randn(20, 20, 20).astype(dtype))
+                stack.append([[A[5:10, :, :], A.T[11:16, :, :]], [A[:5, :, :], A[15:20, :, :]]])
+                stack.append([[[[A]]]])
+
+                for arrays in stack:
+                    self.assertEqual(torch.tensor(arrays, device=device), torch.from_numpy(np.stack(arrays)),
+                                     exact_dtype=True, rtol=1e-05, atol=1e-08)
+
+        # Based on Natalia Gimelshein's test case
+        stack = []
+        A = np.random.rand(2, 3, 4)[:, :, 0]
+        stack.append([A for _ in range(10)])
+        A = np.random.rand(2, 3, 4)[:, 0, 0]
+        stack.append([A for _ in range(10)])
+        A = np.random.rand(2, 3, 4)[:, 0, :]
+        stack.append([A for _ in range(10)])
+
+        for arrays in stack:
```

Comment:
Perhaps we can add a keyword argument for controlling whether the output must be C-contiguous and set its default value to be true? Currently, it is guaranteed to produce a C-contiguous tensor only if the majority of the input numpy arrays are not F-contiguous. A trivial modification can enforce a behaviour that under all cases, C-contiguous tensors will be generated at the cost of a runtime overhead whenever that keyword argument is true.

Reply:
Let's not add a kwarg (because changing the signature is a little risky) and just preserve the current behavior for now. We can revisit changing behavior or extending functionality in the future. This PR is focused on improving performance.